src/AWSCredentials.jl

```diff
@@ -24,75 +27,114 @@ The fields `access_key_id` and `secret_key` hold the access keys used to authent
 """
 mutable struct AWSCredentials
     access_key_id::String
     secret_key::String
     token::String
     user_arn::String
     account_number::String
+    expiry::DateTime
 
-    function AWSCredentials(access_key_id, secret_key,
-                            token="", user_arn="", account_number="")
-        new(access_key_id, secret_key, token, user_arn, account_number)
+    function AWSCredentials(access_key_id,secret_key,
+                            token="", user_arn="", account_number="";
+                            expiry=now(UTC) + DEFAULT_DURATION)
+        new(access_key_id, secret_key, token, user_arn, account_number, expiry)
     end
 end
 
-function Base.show(io::IO,c::AWSCredentials)
-    println(io, string(c.user_arn,
-                       c.user_arn == "" ? "" : " ",
-                       "(",
-                       c.account_number,
-                       c.account_number == "" ? "" : ", ",
-                       c.access_key_id,
-                       c.secret_key == "" ? "" : ", $(c.secret_key[1:3])...",
-                       c.token == "" ? "" : ", $(c.token[1:3])..."),
-                       ")")
-end
+"""
+Holds an [`AWSCredentials`](@ref) object and ensures it is renewed before it can expire.
+Use [`get_credentials`](@ref) to access the credentials.
 
-function Base.copyto!(dest::AWSCredentials, src::AWSCredentials)
-    for f in fieldnames(typeof(dest))
-        setfield!(dest, f, getfield(src, f))
-    end
+The `RenewableAWSCredentials()` constructor tries to load local Credentials from
+environment variables, `~/.aws/credentials`, `~/.aws/config` or EC2 instance credentials.
+To specify the profile to use from `~/.aws/credentials`, do, for example, `RenewableAWSCredentials(profile="profile-name")`.
+"""
+struct RenewableAWSCredentials
+    creds::AWSCredentials
+    renew::Function
+    args::Tuple
 end
-import Base: copy!
-Base.@deprecate copy!(dest::AWSCredentials, src::AWSCredentials) copyto!(dest, src)
 
-function AWSCredentials(;profile=nothing)
+@deprecate AWSCredentials(;profile=nothing) get_credentials(RenewableAWSCredentials(profile=profile))
 
+function RenewableAWSCredentials(;profile=nothing)
+    args = ()
     if haskey(ENV, "AWS_ACCESS_KEY_ID")
 
-        creds = env_instance_credentials()
+        renew = env_instance_credentials
 
     elseif isfile(dot_aws_credentials_file()) || isfile(dot_aws_config_file())
 
-        creds = dot_aws_credentials(profile)
+        renew = dot_aws_credentials
+        args = (profile,)
 
     elseif haskey(ENV, "AWS_CONTAINER_CREDENTIALS_RELATIVE_URI")
 
-        creds = ecs_instance_credentials()
+        renew = ecs_instance_credentials
 
     elseif localhost_maybe_ec2()
 
-        creds = ec2_instance_credentials()
+        renew = ec2_instance_credentials
 
     else
         error("Can't find AWS credentials!")
     end
 
+    creds = renew(args...)
```

Comment:
Are every way of getting credentials renewable? I imagine there are some that are not. Also maybe on renewal we should look at all of various way we can get credentials rather than just the one that worked the first time?

Reply:
The whole point is to not suddenly switch credentials without user action.